### PR TITLE
Use the `referencing` library where available

### PR DESCRIPTION
**Pull request by martinhoyer**
Use the `referencing` library where available instead of deprecated `jsonschema.RefResolver`.

Addressing https://github.com/teemtee/fmf/issues/194 .

Getting rid of deprecation warnings.

### CODE REVIEW

File: fmf.spec [hunk 1]
```diff
@@ -34,6 +34,7 @@ BuildRequires: python%{python3_pkgversion}-jsonschema
 BuildRequires: git-core
 %{?python_provide:%python_provide python%{python3_pkgversion}-%{name}}
 Requires:       git-core
+Recommends:     python%{python3_pkgversion}-referencing
```
Comment (LecrisUT) [inline fmf.spec:37]: I don't think we need it. The dependency is part of `jsonschema` as early as 4.18 [^1] and for rhel case we can just assume it does not have the interface.

[^1]: https://github.com/python-jsonschema/jsonschema/commit/69f3899c0770472f19898d80aea2b7a51fd5ec6b

File: fmf/_compat_jsonschema.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+def create_validator(
+        schema: Dict[str, Any],
+        schema_store: Optional[Dict[str, Any]] = None
+        ) -> Union[jsonschema.validators.Validator, jsonschema.validators.Draft4Validator]:
+    """Create a validator instance based on available jsonschema version."""
+    schema_store = schema_store or {}
+
+    if MODERN_JSONSCHEMA:
+        # Modern approach with referencing
+        registry: Registry = Registry().with_resources([
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:34]:
```suggestion
        registry = Registry().with_resources([
```

Should not be necessary

File: fmf/_compat_jsonschema.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+import jsonschema
+
+try:
+    from jsonschema import validators
+    from referencing import Registry, create
+    from referencing.jsonschema import DRAFT4
+    MODERN_JSONSCHEMA = True
+except ImportError:
+    MODERN_JSONSCHEMA = False
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:16]: Could we make the test be on the `importlib` package version instead? It is less clear what `MODERN_JSONSCHEMA` is referring to otherwise. As for the `create_validator`, what about adding the imports in there?

File: fmf/_compat_jsonschema.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
+"""Compatibility layer for jsonschema validation."""
+
+from __future__ import annotations
+
+import warnings
+from typing import Any, Dict, List, NamedTuple, Optional, Union
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:6]: Let's not mix `__future__.annotation` and `typing.Dict` for now. We should take a hammer at it eventually though.

File: fmf/_compat_jsonschema.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    from referencing.jsonschema import DRAFT4
+    MODERN_JSONSCHEMA = True
+except ImportError:
+    MODERN_JSONSCHEMA = False
+
+
+class JsonSchemaValidationResult(NamedTuple):
+    """Represents JSON Schema validation result."""
+    result: bool
+    errors: List[Any]
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:22]: I don't think these should be in a `_compat` package. Maybe `utils.jsonschema` instead. For this PR though, we can keep them in the original place.

File: fmf/_compat_jsonschema.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        registry: Registry = Registry().with_resources([
+            (uri, create({'$schema': 'http://json-schema.org/draft-04/schema#', **contents}))
+            for uri, contents in schema_store.items()
+            ])
+        cls = validators.create(meta_schema=DRAFT4, validators={})
+        return cls(schema=schema, registry=registry)
+    else:
+        # Legacy approach with RefResolver
+        with warnings.catch_warnings():
+            warnings.filterwarnings('ignore', category=DeprecationWarning)
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:43]: Let's try without the `catch_warnings`. There should not be the DeprecationWarning being triggered if the `MODERN_JSONSCHEMA` is working appropriately, and if it does, we should investigate why it does.

File: fmf/_compat_jsonschema.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    """Represents JSON Schema validation result."""
+    result: bool
+    errors: List[Any]
+
+
+def create_validator(
+        schema: Dict[str, Any],
+        schema_store: Optional[Dict[str, Any]] = None
+        ) -> Union[jsonschema.validators.Validator, jsonschema.validators.Draft4Validator]:
+    """Create a validator instance based on available jsonschema version."""
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:29]: I am more used to `create` naming convention being a static/classmethod. Just a thought if it should be `get` or `create` or something else.

File: fmf/utils.py [hunk 8]
```diff
@@ -10,14 +10,17 @@
...
-import jsonschema
 from filelock import FileLock, Timeout
 from ruamel.yaml import YAML, scalarstring
 from ruamel.yaml.comments import CommentedMap
 
 import fmf.base
+# F401 utils is entry point, while _compat is being used
+# isort with autopep8 made it to two lines. Ruff needed.
+from fmf._compat_jsonschema import JsonSchemaValidationResult  # noqa: F401
+from fmf._compat_jsonschema import validate_data  # noqa: F401
```
Comment (LecrisUT) [inline fmf/utils.py:23]: Why didn't `fmf._compat.jsonschema` work?

File: fmf/utils.py [hunk 9]
```diff
@@ -419,7 +422,7 @@ class Logging:
     _level = LOG_WARN
 
     # Already initialized loggers by their name
-    _loggers = dict()
+    _loggers: dict[str, logging.Logger] = dict()
```
Comment (LecrisUT) [inline fmf/utils.py:425]: This doesn't have `__future__.annotations` here so it should be using `Dict` instead here. But also let's deffer the change.

**Review by LecrisUT — CHANGES_REQUESTED** [review]

File: fmf.spec (hunk 1 above)
Comment (martinhoyer) [reply fmf.spec:37]: oh cool, thanks. Removing.

File: fmf/utils.py (hunk 9 above)
Comment (martinhoyer) [reply fmf/utils.py:425]: I think `dict` is already in Python 3.9, no?

Comment (happz) [reply fmf/utils.py:425]: Correct, Python 3.9 allows builtin types as annotations.

**Comment by happz** [comment]
So many distinct changes squashed into a single PR :(

**Comment by martinhoyer** [comment]
> So many distinct changes squashed into a single PR :(

Well, I couldn't run pre-commit without it and didn't want to make workarounds.

**Comment by martinhoyer** [comment]
...but fwiw, for there is a pr for the last commit https://github.com/teemtee/fmf/pull/258

Context for the autopep8 is that the previous hook repo was unmaintained and stopped being updated (doesn't work at all on py3.13). Replaced it with what was recommended as replacement and with `--strict`, the `is_instance` changes were needed to pass that check.

File: fmf/_compat_jsonschema.py (hunk 3 above)
Comment (martinhoyer) [reply fmf/_compat_jsonschema.py]: How about now, without the flag?

File: fmf/_compat_jsonschema.py (hunk 5 above)
Comment (martinhoyer) [reply fmf/_compat_jsonschema.py]: moved back, thanks

File: fmf/_compat_jsonschema.py (hunk 6 above)
Comment (martinhoyer) [reply fmf/_compat_jsonschema.py]: True. Removed.

File: fmf/_compat_jsonschema.py (hunk 7 above)
Comment (martinhoyer) [reply fmf/_compat_jsonschema.py]: renamed

File: fmf/utils.py (hunk 8 above)
Comment (martinhoyer) [reply fmf/utils.py]: hmm, I don't remember. Any suggestion about how the structure should look like are welcome.

File: fmf/utils.py (hunk 9 above)
Comment (LecrisUT) [reply fmf/utils.py:425]: Ah right, must have confused it with `Union` compatibility issue

File: fmf/utils.py (hunk 8 above)
Comment (LecrisUT) [reply fmf/utils.py]: The design I saw was have a `_compat` package and under that the same structure as the most recent interface that will eventually replace it, e.g. `_compat.collections.abc` for `Itterator`, `_compat.warnings` for `deprecated` etc. That way the eventual change is to simply drop the `_compat` from the `from import`.

The part that I didn't see used was something like `from _compat import warnings` that would cover both compatibilities like `warnings.deprecated` and normal functions like `warnings.warn`. It seems sensible to me if you want to eventually use the fully-qualified naming, but not sure if there are pitfalls that I'm overlooking.

File: fmf/utils.py [hunk 10]
```diff
@@ -955,6 +957,7 @@ def validate_data(data, schema, schema_store=None):
             jsonschema.exceptions.RefResolutionError,
             jsonschema.exceptions.UnknownType
             ) as error:
+        from fmf.utils import JsonSchemaError
```
Comment (LecrisUT) [inline fmf/utils.py:960]: Unneeded? We are in `utils.py`

File: fmf/utils.py [hunk 11]
```diff
@@ -912,34 +915,33 @@ def dict_to_yaml(data, width=None, sort=False):
...
+        data: The data to validate
+        schema: The JSON Schema to validate against
+        schema_store: Optional dict of schema references and their content
 
-        result ... boolean representing the validation result
-        errors ... A list of validation errors
+    Returns:
+        JsonSchemaValidationResult with:
+            result: boolean representing the validation result
+            errors: A list of validation errors
```
Comment (LecrisUT) [inline fmf/utils.py:939]: What docstyle has been adopted?

File: fmf/_compat_jsonschema.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+
+def get_validator(
+        schema: Any,
+        schema_store: Optional[dict[str, Any]] = None
+        ) -> Draft4Validator:
+    """Create a validator instance based on available jsonschema version."""
+    # Validate schema is a dict
+    if not isinstance(schema, dict):
+        from fmf.utils import JsonSchemaError
+        raise JsonSchemaError(f'Invalid schema type: {type(schema)}. Schema must be a dictionary.')
```
Comment (LecrisUT) [inline fmf/_compat_jsonschema.py:17]: Probably this check can be removed when `mypy` is added. Should we add a todo note for that?

File: fmf/_compat_jsonschema.py (hunk 3 above)
Comment (LecrisUT) [reply fmf/_compat_jsonschema.py]: 👍

**Comment by psss** [comment]
> > So many distinct changes squashed into a single PR :(
> 
> Well, I couldn't run pre-commit without it and didn't want to make workarounds.

Let's address at least the `pre-commit` changes separately in #259?

File: fmf/utils.py (hunk 11 above)
Comment (martinhoyer) [reply fmf/utils.py:939]: hah, you have to ask my IDE, I'm just pressing TAB :D  
Joking aside, I do rely on formatters and linters for this stuff. :/

Comment (martinhoyer) [reply fmf/utils.py:939]: @LecrisUT maybe we can move https://github.com/teemtee/fmf/pull/240 forward?

File: fmf/utils.py (hunk 10 above)
Comment (LecrisUT) [reply fmf/utils.py:947]: You forgot this one in the recent rebase?

File: fmf/utils.py (hunk 8 above)
Comment (LecrisUT) [reply fmf/utils.py]: Lgtm, don't think we can follow the other comoat designs here, but at least the change is localized so it helps to be in this form 👍

**Review by LecrisUT — APPROVED** [review]

File: fmf/utils.py (hunk 10 above)
Comment (martinhoyer) [reply fmf/utils.py:947]: Huh, I specifically checked for this. Must be going insane. Thanks mate.

File: .gitignore [hunk 13]
```diff
@@ -9,4 +9,77 @@ dist
...
+parts/
+sdist/
+var/
+wheels/
+share/python-wheels/
+*.egg-info/
+.installed.cfg
+*.egg
+MANIFEST
+tmt.1
```
Comment (psss) [inline .gitignore:39]: Irrelevant copy & paste here? Should be `fmf.1` instead?

File: .pre-commit-config.yaml [hunk 14]
```diff
@@ -9,7 +9,6 @@ repos:
           - --recursive
           - --in-place
           - --aggressive
-          - --aggressive
```
Comment (psss) [inline .pre-commit-config.yaml:12]: Why this change?

File: fmf/utils.py [hunk 15]
```diff
@@ -912,34 +913,22 @@ def dict_to_yaml(data, width=None, sort=False):
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 class JsonSchemaValidationResult(NamedTuple):
-    """ Represents JSON Schema validation result """
-
+    """Represents JSON Schema validation result."""
```
Comment (psss) [inline fmf/utils.py:916]: Can we keep these as they are until there's a concensus on the formatting?

**Review by psss — CHANGES_REQUESTED** [review]
Thanks for looking into this! Looks good, just a few minor comments.

File: .gitignore (hunk 13 above)
Comment (martinhoyer) [reply .gitignore]: thanks

File: .pre-commit-config.yaml (hunk 14 above)
Comment (martinhoyer) [reply .pre-commit-config.yaml]: it's back. Thanks.

File: fmf/utils.py (hunk 15 above)
Comment (martinhoyer) [reply fmf/utils.py]: reverted in fixup.

File: fmf/utils.py (hunk 11 above)
Comment (psss) [reply fmf/utils.py]: > @LecrisUT maybe we can move #240 forward?

Yeah, let's enable it in `1.6`.

**Review by psss — APPROVED** [review]
